test/library/packages/DistributedBag/nqueens.chpl

```diff
@@ -0,0 +1,232 @@
+
+  // Root-initializer
+  proc init(const N: int) {
+    this.dom = {0..#N};
+    this.board = 0..#N;
+  }
+
+  // Copy-initializer
+  proc init(other: Node) {
+    this.dom   = other.dom;
+    this.board = other.board;
+    this.depth = other.depth;
+  }
+}
+
+
+/*** Backtracking operators ***/
+
+// Check if the position is safe, i.e., no two queens share the same diagonals.
+proc isSafe(const board: [] int, const queen_num: int, const row_pos: int): bool
+{
+  // For each queen before this one
+  for i in 0..#queen_num {
+    // Get the row position
+    const other_row_pos = board[i];
+
+    // Check diagonals
+    if (other_row_pos == row_pos - (queen_num - i) ||
+        other_row_pos == row_pos + (queen_num - i)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Decompose a node into a set of children nodes.
+proc decompose(const parent: Node, const N: int, ref tree_loc: int,
+  ref num_sol: int): list(Node)
+{
+  var children: list(Node);
+
+  const depth = parent.depth;
+
+  if (depth == N) then num_sol += 1; // all queens are placed
+  else {
+    for j in depth..N-1 {
+      if isSafe(parent.board, depth, parent.board[j]) {
+        var child = new Node(parent);
+        child.board[depth] <=> child.board[j];
+        child.depth += 1;
+        children.pushBack(child);
+        tree_loc += 1;
+      }
+    }
+  }
+
+  return children;
+}
+
+// Check termination condition
+inline proc allIdle(const arr: [] atomic bool): bool
+{
+  for elt in arr do
+    if (elt.read() == BUSY) then return false;
+
+  return true;
+}
+
+
+/*** Distributed Search ***/
+
+proc nqueens_search(const N: int)
+{
+  // Global variables (synchronization, termination)
+  const PrivateSpace: domain(1) dmapped privateDist();
+  var eachLocaleState: [PrivateSpace] atomic bool = BUSY;
+  allLocalesBarrier.reset(here.maxTaskPar);
+
+  // Statistics
+  var eachExploredTree: [PrivateSpace] int;
+  var eachExploredSol: [PrivateSpace] int;
+
+  // Initialization
+  var bag = new distBag(Node, targetLocales = Locales);
+  var root = new Node(N);
+  bag.add(root, 0);
+
+  // Parallel search
+  coforall loc in Locales with (ref eachExploredTree, ref eachExploredSol,
+    ref eachLocaleState) do on loc {
+
+    // Local variables
+    var eachTaskState: [0..#here.maxTaskPar] atomic bool = BUSY;
+
+    // Local statistics
+    var eachLocalExploredTree: [0..#here.maxTaskPar] int;
+    var eachLocalExploredSol: [0..#here.maxTaskPar] int;
+
+    coforall taskId in 0 .. #here.maxTaskPar with (ref eachLocalExploredTree,
+      ref eachLocalExploredSol, ref eachTaskState, ref eachLocaleState) {
+
+      // Task variables
+      ref tree_loc = eachLocalExploredTree[taskId];
+      ref num_sol = eachLocalExploredSol[taskId];
+
+      allLocalesBarrier.barrier();
+
+      while true {
+
+        // Try to get an element
+        var (hasWork, parent) = bag.remove(taskId);
+
+        // Check termination
+        if hasWork {
+          eachTaskState[taskId].write(BUSY);
```

Comment:
It might improve performance to write `ref curTaskState = eachTaskState[taskId]` just outside of the `while true` so that the array lookup does not need to be re-done each time through here. Same for `eachLocaleState`. Of course, this is not worth bothering with if you don't care about this performance of this test.

Reply:
This test doesn't deal with performance, but any minor modification that can make the code more optimized is welcomed.